scantool/scantool_850/parser.py

```diff
@@ -0,0 +1,204 @@
+import sys
+import re
+
+# a class representing an OBD-II Diagnostic Trouble Code (DTC)
+class dtc:
+    # constructor
+    def __init__(self, ecu, code, prefix, suffix, description):
+        self.code = code
+        self.description = description
+        self.ecu = ecu
+        self.prefix = prefix
+        self.suffix = suffix
+
+# a class representing an ecu address and description
+class ecu:
+    # constructor
+    def __init__(self, address, description, prefix):
+        self.address = address
+        self.description = description
+        self.prefix = prefix
+    
+
+# this loads dtcs from the xiaotec file export
+def load_dtcs(file_path):
+    dtcs = []
+    ecus = []
+
+    with open(file_path, 'r') as file:
+        current_ecu = None
+        strings = {}
+        for line in file:
+            # regular expression to match 01 - description
+            match = re.match(r'([0-9A-F][0-9A-F]) - (.+)', line.strip())
+            if match:
+                current_ecu = ecu(match.groups()[0], match.groups()[1], 'unknown')
+                ecus.append(current_ecu)
+                # print(current_ecu.address)
+            
+            # regular expression to match dtc.contains("<code>") once
+            match = re.match(r'^if \(dtc.contains\("([0-9A-F]+)"\)\) ?// ?([^\-]+)\-?([0-9]+) ?\(?([a-z])?\)?', line.strip())
+            if match:
+                string_key = match.groups()[2]
+                if match.groups()[3]:
+                    string_key += match.groups()[3]
+                dtcs.append(dtc(current_ecu.address, match.groups()[0], match.groups()[1], match.groups()[2].strip(), string_key))
+                current_ecu.prefix = match.groups()[1]
+
+            # regular expression to match dtc.contains("<code>") twice
+            match = re.match(r'^if \(dtc.contains\("([0-9A-F]+)"\) ?\|\| ?dtc.contains\("([0-9A-F]+)"\)\) ?// ?([^\-]+)\-?([0-9]+) ?\(?([a-z])?\)?', line.strip())
+            if match:
+                string_key = match.groups()[2]
+                if match.groups()[3]:
+                    string_key += match.groups()[3]
+                dtcs.append(dtc(current_ecu.address, match.groups()[0], match.groups()[2], match.groups()[3], string_key))
+                dtcs.append(dtc(current_ecu.address, match.groups()[1], match.groups()[2], match.groups()[3], string_key))
+            
+            match = re.match(r'<string name="([0-9A-Za-z_]+)">(.+)</string>', line.strip())
+            if match:
+                if not current_ecu.address in strings:
+                    strings[current_ecu.address] = {}
+                strings[current_ecu.address][match.groups()[0]] = match.groups()[1]
+    
+    for d in dtcs:
+        key = d.prefix + d.suffix
+        ecu_key = d.ecu
+        if ecu_key == '2E':
```

Comment:
wha'ts special about 2E / 2F here ?

Reply:
2E is the left power seat and 2F is the right power seat. The DTCs are the same, only the address is different. (Except on model year '96, where they put them both on 2F, and the tester can't talk to them unless you unhook one or the other. Oops.)